api/model/src/main/java/org/projectnessie/model/Util.java

```diff
@@ -61,9 +124,39 @@ public static List<String> fromPathString(String encoded) {
    * @return String encoded for path use.
    */
   public static String toPathString(List<String> elements) {
-    return elements.stream()
-        .map(x -> x.replace(DOT, GROUP_SEPARATOR).replace(ZERO_BYTE, GROUP_SEPARATOR))
-        .collect(Collectors.joining("."));
+    StringBuilder sb = new StringBuilder();
+    for (String element : elements) {
+      if (sb.length() > 0) {
+        sb.append('.');
+      }
+      int l = element.length();
+      for (int i = 0; i < l; i++) {
+        char c = element.charAt(i);
+        sb.append(c == DOT || c == ZERO_BYTE ? GROUP_SEPARATOR : c);
+      }
+    }
+    return sb.toString();
+  }
+
+  public static String toPathStringEscaped(List<String> elements) {
+    StringBuilder sb = new StringBuilder();
+    for (String element : elements) {
+      if (sb.length() > 0) {
+        sb.append('.');
+      }
+      int l = element.length();
+      for (int i = 0; i < l; i++) {
+        char c = element.charAt(i);
+        if (c == DOT) {
+          sb.append("..~");
+        } else if (c == '/') {
```

Comment:
I believe we also need to handle `\` (example value `hellow\world`) and `%` (ex: `%hellow`). They cause URI violations.

Reply:
suggestion: since we have multiple escape cases, maybe add a generic translation table for the char following `..`?

Reply:
2 * (512 bytes + array header) for 4 cases feels overkill